crossplane/crossplane
# Add `Usage` type for Deletion Ordering and Resource Protection

### Description of your changes

This PR implements the `Usage` type proposed in [this one-pager](https://github.com/crossplane/crossplane/blob/master/design/one-pager-generic-usage-type.md) as an alpha feature.

This would enable ordered deletions when there is a usage dependency between composed resources. A typical example is having a helm `Release` together with a GKE `Cluster` resource where the release is installed into the cluster. Currently, we end up orphaned `Release` resources when the `Cluster` is gone before the release is properly uninstalled. With this PR, it would be possible to include the following `Usage` resource in the same composition to prevent the `Cluster` deletion before the `Release` is gone. 

```yaml
  resources:
    - name: cluster
      base:
        apiVersion: container.gcp.upbound.io/v1beta1
        kind: Cluster
        ...
    - name: release
      base:
        apiVersion: helm.crossplane.io/v1beta1
        kind: Release
        ...
  - name: release-uses-cluster
    base:
      apiVersion: apiextensions.crossplane.io/v1alpha1
      kind: Usage
      spec:
        of:
          apiVersion: container.gcp.upbound.io/v1beta1
          kind: Cluster
          resourceSelector:
            matchControllerRef: true
        by:
          apiVersion: helm.crossplane.io/v1beta1
          kind: Release
          resourceSelector:
            matchControllerRef: true
```

Fixes #3393 

I have:

- [x] Added or updated unit **and** E2E tests for my change.
- [x] Opened a PR updating the [docs](https://docs.crossplane.io/contribute/contribute/), if necessary -> https://github.com/crossplane/docs/pull/532

Todo's:

- [x] Fix todo's in `UsageComposition` e2e tests by adding more assess's
- [x] Follow up with other possible solutions to the garbage collector [backoff problem](https://github.com/crossplane/crossplane/pull/4215#discussion_r1279174880) (I'll create a ticket) 

### How to Test

I've built a custom image from this PR. You can install and enable the feature using the following command:

```shell
helm repo add crossplane-master https://charts.crossplane.io/master --force-update
helm upgrade --install crossplane --namespace crossplane-system crossplane-master/crossplane --version v1.14.0-rc.0.190.g78bdab10 --create-namespace --set image.repository=turkenh/crossplane --set image.tag=v1.14.0-rc.0.198.g9049ab9c --set "args={--debug,--enable-usages}"
```

You can also check the examples used in end to end testing under the directory: `test/e2e/manifests/apiextensions/usage`

```shell
❯ cd test/e2e/manifests/apiextensions/usage/standalone/
❯ kubectl apply -f setup/
provider.pkg.crossplane.io/provider-nop created
❯ kubectl wait provider.pkg provider-nop --for condition=healthy --timeout 2m
provider.pkg.crossplane.io/provider-nop condition met
❯ kubectl apply -f with-by/
usage.apiextensions.crossplane.io/using-uses-used created
nopresource.nop.crossplane.io/used-resource created
nopresource.nop.crossplane.io/using-resource created
❯ kubectl apply -f with-reason/
usage.apiextensions.crossplane.io/protect-a-resource created
nopresource.nop.crossplane.io/protected-resource created
❯ kubectl get usages
NAME                 DETAILS                                                     READY   AGE
protect-a-resource   This resource is protected!                                 True    15s
using-uses-used      NopResource/using-resource uses NopResource/used-resource   True    19s
❯ kubectl delete -f with-by/used.yaml
Error from server (This resource is in-use by 1 Usage(s), including the Usage "using-uses-used" by resource NopResource/using-resource.): error when deleting "with-by/used.yaml": admission webhook "nousages.apiextensions.crossplane.io" denied the request: This resource is in-use by 1 Usage(s), including the Usage "using-uses-used" by resource NopResource/using-resource.
❯ kubectl delete -f with-reason/used.yaml
Error from server (This resource is in-use by 1 Usage(s), including the Usage "protect-a-resource" with reason: "This resource is protected!".): error when deleting "with-reason/used.yaml": admission webhook "nousages.apiextensions.crossplane.io" denied the request: This resource is in-use by 1 Usage(s), including the Usage "protect-a-resource" with reason: "This resource is protected!".
```

[contribution process]: https://git.io/fj2m9
